app/controllers/topics_controller.rb

```diff
@@ -150,7 +150,7 @@ def create
         :attachments => params[:topic][:posts_attributes]["0"][:attachments])
     end
 
-    @topic.errors.add(:attachments, "Not valid format") unless @post.errors.empty?
+    @topic.validate_attachment_format(@post.errors)
```

Comment:
This is better, but is still calling it inline in the controller, rather than in the validates callback of the model itself.  You can basically take this same method and apply it as a custom validator: http://guides.rubyonrails.org/active_record_validations.html#performing-custom-validations